src/api/common/connection.js

```diff
@@ -0,0 +1,125 @@
+'use strict';
+const {EventEmitter} = require('events');
+const WebSocket = require('ws');
+
+function createDisconnectionError(id) {
+  return {id: id, status: 'error', type: 'error', error: 'disconnected'};
+}
+
+function isStreamMessageType(type) {
+  return type === 'ledgerClosed' ||
+         type === 'transaction' ||
+         type === 'path_find';
+}
+
+class Connection extends EventEmitter {
+  constructor(url) {
+    super();
+    this._url = url;
+    this._ws = null;
+    this._nextRequestID = 1;
+  }
+
+  _onMessage(message) {
+    try {
+      const data = JSON.parse(message);
+      if (data.type === 'response') {
+        if (!(Number.isInteger(data.id) && data.id >= 0)) {
+          throw new Error('valid id not found in response');
+        }
+        this.emit(data.id.toString(), data);
+      } else if (isStreamMessageType(data.type)) {
+        this.emit(data.type, data);
+      } else {
+        throw new Error('unrecognized message type: ' + data.type);
+      }
+    } catch (error) {
+      this.emit('badMessage', error);
+    }
+  }
+
+  state() {
+    return this._ws ? this._ws.readyState : WebSocket.CLOSED;
+  }
+
+  connect() {
+    return new Promise((resolve) => {
+      if (this.state() === WebSocket.OPEN) {
+        resolve();
+      } else if (this.state() === WebSocket.CONNECTING) {
+        this._ws.once('open', resolve);
+      } else {
+        this._ws = new WebSocket(this._url);
+        this._ws.on('message', this._onMessage.bind(this));
+        this._ws.once('close', () => this.connect().then());
+        this._ws.once('open', resolve);
+      }
+    });
+  }
+
+  disconnect() {
+    return new Promise((resolve) => {
+      if (this.state() === WebSocket.CLOSED) {
+        resolve();
+      } else if (this.state() === WebSocket.CLOSING) {
+        this._ws.once('close', resolve);
+      } else {
+        this._ws.removeAllListener('close');
+        this._ws.once('close', resolve);
+        this._ws.close();
+      }
+    });
+  }
+
+  reconnect() {
+    return this.disconnect().then(() => this.connect());
+  }
+
+  _send(message) {
+    this._ws.send(message);
+  }
+
+  _sendWhenReady(message) {
+    return new Promise((resolve) => {
+      if (this.state() === WebSocket.OPEN) {
+        resolve(this._send(message));
+      } else {
+        this.once('open', () => resolve(this._send(message)));
```

Comment:
But here `send` will be executed in context of `open` event handler, not
inside of promise, no?
19 жовт. 2015 20:31 "Chris Clark" <notifications@github.com> пише:

> In src/api/common/connection.js
> <https://github.com/ripple/ripple-lib/pull/601#discussion_r42400596>:
>
> > +  }
> > +
> > +  reconnect() {
> > +    return this.disconnect().then(() => this.connect());
> > +  }
> > +
> > +  _send(message) {
> > +    this._ws.send(message);
> > +  }
> > +
> > +  _sendWhenReady(message) {
> > +    return new Promise((resolve) => {
> > +      if (this.state() === WebSocket.OPEN) {
> > +        resolve(this._send(message));
> > +      } else {
> > +        this.once('open', () => resolve(this._send(message)));
>
> Yeah, should be this._ws.once('open', ...
>
> Promises capture exceptions, so if send throws an error, it will be
> caught by the catch in request().
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ripple/ripple-lib/pull/601/files#r42400596>.
>